torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,215 @@
+    std::stack<Block*> blocks({graph->block()});
+    recordMutableAttrs(graph, module_clone);
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+        Node* n = *it;
+        it++; // advance iterator bc the current node may be destroyed
+
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          WithInsertPoint guard(n);
+          auto name = n->s(attr::name);
+          auto attr_module = module_clone;
+          auto input = n->inputs()[0]->node();
+          if (!findConstantAttr(input, name, attr_module)) {
+            continue;
+          }
+          auto attr = attr_module.attr(name);
+          if (attr.type()->cast<ClassType>()) {
+            continue;
+          }
+
+          if (attr.isTensor()) {
+            auto t = attr.toTensor();
+            t.set_requires_grad(false);
+          }
+          // FIXME: support Dict, TUPLE, genericList and DoubleList constants
+          // to resolve GetAttr.
+          if (attr.isTuple() || attr.isGenericList() || attr.isGenericDict() ||
+              attr.isDoubleList()) {
+            // TORCH_CHECK(false, "Constant Dict, GenericList, Tuple, and
+            // DoubleList  are not yet supported");
+            continue;
+          }
+          Value* param_const = graph->insertConstant(attr);
+          n->outputs().at(0)->replaceAllUsesWith(param_const);
+          n->removeAllInputs();
+          n->destroy();
+        }
+      }
+    }
+  }
+
+  // This function cleans up  the Frozen module. it performs the following:
+  // 1) Remove unused all attributes.
+  // 2) Remove all unreferenced submodules
+  // 3) Remove non pulic unreferenced methods.
+  // TODO: do #3 because there is no API to 'unsafely' remove methods.
+
+  //  This functions relies on unsafeRemoveAttr.
+  void cleanupFrozenModule(
+      std::shared_ptr<Graph>& graph,
+      script::Module& module_clone) {
+    std::vector<std::string> attrsToRemove;
+    auto type = module_clone.type();
+    size_t N = type->numAttributes();
+    auto KeepAttrs = getReferencedAttrs(graph, module_clone);
+    for (size_t i = 0; i < N; ++i) {
+      auto attrTy = type->getAttribute(i);
+      auto name = type->getAttributeName(i);
+      if (!KeepAttrs.count(name)) {
+        attrsToRemove.push_back(name);
+      }
+    }
+    for (auto name : attrsToRemove) {
+      module_clone._ivalue()->unsafeRemoveAttr(name);
+      module_clone.type()->unsafeRemoveAttribute(name);
+    }
+  }
+}; // class AttributePropagator
+} // namespace
+
+script::Module FreezeModule(const script::Module& module) {
+  AttributePropagator attrPropagator;
+  auto module_clone = module.clone();
+  script::Method method = module_clone.get_method("forward");
+  auto graph = method.graph();
+  Inline(*graph);
+  runOptimization(graph);
```

Comment:
Should we be unrolling loops twice here?

Reply:
Good catch Elias!